dnf/transaction_sr.py

```diff
@@ -0,0 +1,600 @@
+        :param errors: a list of error classes or a string with an error description
+        """
+
+        if isinstance(errors, (list, tuple)):
+            if len(errors) > 1:
+                msg = _('Errors in "{filename}":').format(filename=filename)
+                for error in errors:
+                    msg += "\n  " + str(error)
+
+                super(TransactionFileError, self).__init__(msg)
+                return
+
+            else:
+                errors = str(errors[0])
+
+        msg = _('Error in "{filename}": {error}').format(filename=filename, error=errors)
+        super(TransactionFileError, self).__init__(msg)
+
+
+class IncompatibleTransactionVersionError(TransactionFileError):
+    def __init__(self, filename, msg):
+        super(IncompatibleTransactionVersionError, self).__init__(filename, msg)
+
+
+def _check_version(version, filename):
+    major, minor = version.split('.')
+
+    try:
+        major = int(major)
+    except ValueError as e:
+        raise TransactionFileError(
+            filename,
+            _('Invalid major version "{major}", number expected.').format(major=major)
+        )
+
+    try:
+        int(minor)  # minor is unused, just check it's a number
+    except ValueError as e:
+        raise TransactionFileError(
+            filename,
+            _('Invalid minor version "{minor}", number expected.').format(minor=minor)
+        )
+
+    if major != VERSION_MAJOR:
+        raise IncompatibleTransactionVersionError(
+            filename,
+            _('Incompatible major version "{major}", supported major version is "{major_supp}".')
+                .format(major=major, major_supp=VERSION_MAJOR)
+        )
+
+
+def serialize_transaction(transaction):
+    """
+    Serializes a transaction to a data structure that is equivalent to the stored JSON format.
+    :param transaction: the transaction to serialize (an instance of dnf.db.history.TransactionWrapper)
+    """
+
+    data = {
+        "version": VERSION,
+    }
+    rpms = []
+    groups = []
+    environments = []
+
+    for tsi in transaction.packages():
+        if tsi.is_package():
+            rpms.append({
+                "action": tsi.action_name,
+                "nevra": tsi.nevra,
+                "reason": libdnf.transaction.TransactionItemReasonToString(tsi.reason),
+                "repo_id": tsi.from_repo
+            })
+
+        elif tsi.is_group():
+            group = tsi.get_group()
+
+            group_data = {
+                "action": tsi.action_name,
+                "id": group.getGroupId(),
+                "packages": [],
+                "package_types": libdnf.transaction.compsPackageTypeToString(group.getPackageTypes())
+            }
+
+            for pkg in group.getPackages():
+                group_data["packages"].append({
+                    "name": pkg.getName(),
+                    "installed": pkg.getInstalled(),
```

Comment:
Does it mean that pkg.getName() return the name from original metadata and getInstalled() returns true if rpm of given name was also part of the transaction?

Reply:
I assume so.

I've just taken all the data from the SWDB classes, stored them in the transaction and then set them back to SWDB classes during replay. Unless I've missed something this creates the same SQLite records in the DB. The information is not used for anything else.

Reply:
Perfect